File: apps/desktop/src/mocks/cleanup.ts
// The mock's half of `crates/core/src/cleanup/` and of the desktop's `cleanup.rs`: which
// requests a batch may run, the steps the demo module plans for them, and what running them
// does to the mock's world.
//
// Every rule here mirrors one the backend enforces, because this is the only oracle the UI
// tests of the Cleanup screen have. `crates/core/tests/fixtures/cleanup-cases.json` is
// answered by both sides — by `screen` and `reversible` here and in `engine.rs` — so a rule
// that changes in one and not the other reddens a test rather than drifting quietly.

import {
  CLEANUP_PROGRESS_EVENT,
  type ActionSpec,
  type BlockReason,
  type CleanupEntry,
  type CleanupEntryOutcome,
  type CleanupOutcome,
  type CleanupPreview,
  type CleanupPreviews,
  type CleanupRequest,
  type CleanupResult,
  type DeletionMode,
  type EntryResult,
  type EntryStatus,
  type Item,
  type NodeKind,
  type StepEffect,
  type StepView,
} from '../lib/ipc';
import { commandLine } from '../lib/commandLine';
import { appendCleanupToLog } from './actionLog';
import { checkPath, limitsFor } from './actions';
import { FIXTURE_ROOT, fixtureEntry, patchTree, type FixtureNode } from './fixtures';
import {
  DEMO_LAST_CLEANUP,
  DEMO_SANDBOX,
  forgetItems,
  heldItems,
  modulesOf,
  mockModulesRefresh,
  publish,
} from './modules';

/** A path a step deletes, and what the module saw there. */
interface Target {
  path: string;
  kind: NodeKind;
}

/** `Step` of `module/step.rs`, with the command as its argv. */
export type MockStep =
  | { step: 'delete'; target: Target }
  | { step: 'run'; argv: string[]; effect: 'housekeeping' | 'destroys' }
  | { step: 'run'; argv: string[]; effect: 'removes'; target: Target };

/**
 * `screen`: rules 1 to 3 of the design (section 8), in order; the first that applies
 * decides. Disk-free, which is why the shared cases call it directly.
 */
export function screen(
  request: CleanupRequest,
  held: ReadonlyMap<string, Item>,
): { item: Item; action: ActionSpec } | { reason: BlockReason } {
  // 1. Not among the held results.
  const item = held.get(request.item);
  if (item === undefined) {
    return { reason: 'missing' };
  }
  // 2. The item is not what the screen showed: no such action, or no such option.
  const action = item.actions.find((spec) => spec.id === request.action);
  if (action === undefined) {
    return { reason: 'kindChanged' };
  }
  const optionOf = (id: string) => action.options.find((option) => option.id === id);
  if (request.options.some((id) => optionOf(id) === undefined)) {
    return { reason: 'kindChanged' };
  }
  // 3. A Keep verdict is lifted by a force option and by nothing else.
  const forced = request.options.some((id) => optionOf(id)?.force === true);
  if (item.verdict.level === 'keep' && !forced) {
    return { reason: 'kept' };
  }
  return { item, action };
}

/**
 * `reversible`: the mode is Trash, and every step is a deletion or housekeeping. Takes the
 * kinds of the steps, which is all it reads — the shared cases name nothing else.
 */
export function reversible(
  steps: ReadonlyArray<'delete' | StepEffect>,
  mode: DeletionMode,
): boolean {
  return mode === 'trash' && steps.every((kind) => kind === 'delete' || kind === 'housekeeping');
}

/** The kinds `reversible` reads, of the steps a module planned. */
function kindsOf(steps: readonly MockStep[]): Array<'delete' | StepEffect> {
  return steps.map((step) => (step.step === 'delete' ? 'delete' : step.effect));
}

/** `plan` of the demo module: two steps against one for a folder, `rm` for an object. */
function planDemo(item: Item, action: ActionSpec, mode: DeletionMode): MockStep[] {
  if (item.path === null) {
    return [];
  }
  if (item.kind === 'folder' && action.id === 'delete') {
    const remove: MockStep = { step: 'delete', target: { path: item.path, kind: 'dir' } };
    return mode === 'trash'
      ? [remove, { step: 'run', argv: ['touch', DEMO_LAST_CLEANUP], effect: 'housekeeping' }]
      : [remove];
  }
  if (item.kind === 'object' && action.id === 'remove') {
    return [
      {
        step: 'run',
        argv: ['rm', item.path],
        effect: 'removes',
        target: { path: item.path, kind: 'file' },
      },
    ];
  }
  return [];
}

/** The targets a plan deletes. */
function targetsOf(steps: readonly MockStep[]): Target[] {
  return steps.flatMap((step) => ('target' in step ? [step.target] : []));
}

/**
 * Whether a target of the demo's sandbox is on the mock's disk. The demo lives outside the
 * fixture tree, so this asks the demo's own list: the item is held exactly while its entry is.
 */
function onDisk(target: Target, held: ReadonlyMap<string, Item>): boolean {
  if (target.path === DEMO_SANDBOX || target.path.startsWith(`${DEMO_SANDBOX}/`)) {
    return [...held.values()].some((item) => item.path === target.path);
  }
  return fixtureEntry(target.path) !== undefined;
}

/** One request, planned and checked. */
interface Planned {
  request: CleanupRequest;
  item: Item | undefined;
  title: string;
  action: string;
  steps: MockStep[];
  judged: string[];
  size: number;
  status: EntryStatus;
}

/** Rules 1 to 6 over one batch, in one mode. */
function planAll(
  requests: readonly CleanupRequest[],
  held: ReadonlyMap<string, Item>,
  mode: DeletionMode,
): Planned[] {
  const limits = limitsFor(FIXTURE_ROOT);
  const planned = requests.map((request): Planned => {
    const known = held.get(request.item);
    const spec = known?.actions.find((action) => action.id === request.action);
    const entry: Planned = {
      request,
      item: known,
      title: known?.title ?? request.item,
      action: spec?.label ?? request.action,
      steps: [],
      judged: [],
      size: spec?.estimatedFree ?? 0,
      status: { state: 'ready' },
    };
    const screened = screen(request, held);
    if ('reason' in screened) {
      entry.status = { state: 'blocked', reason: screened.reason };
      return entry;
    }
    // 4. The module plans; an empty plan would be reported as done having done nothing.
    entry.steps = planDemo(screened.item, screened.action, mode);
    if (entry.steps.length === 0) {
      entry.status = { state: 'blocked', reason: 'malformed' };
      return entry;
    }
    // 5. Every target through the guards and the disk. The mock's folders and files are
    //    what the module says they are, so the kind comparison has nothing to refuse.
    for (const target of targetsOf(entry.steps)) {
      const checked = checkPath(limits, target.path);
      if ('reason' in checked) {
        entry.status = { state: 'blocked', reason: checked.reason };
        return entry;
      }
      if (!onDisk(target, held)) {
        entry.status = { state: 'blocked', reason: 'missing' };
        return entry;
      }
      entry.judged.push(checked.judged);
    }
    return entry;
  });
  // 6. Across the batch, over the still-ready entries: `drop_nested`'s predicate, comparing
  //    only targets of different entries.
  const ready = planned
    .map((entry, index) => ({ entry, index }))
    .filter(({ entry }) => entry.status.state === 'ready');
  const inside = (path: string, outer: string) => path === outer || path.startsWith(`${outer}/`);
  for (const { entry, index } of ready) {
    const nested = entry.judged.some((target) =>
      ready.some(
        (other) =>
          other.index !== index &&
          other.entry.judged.some(
            (outer) => inside(target, outer) && (outer !== target || other.index < index),
          ),
      ),
    );
    if (nested) {
      entry.status = { state: 'blocked', reason: 'nested' };
    }
  }
  return planned;
}

function view(step: MockStep, mode: DeletionMode): StepView {
  if (step.step === 'delete') {
    return mode === 'trash'
      ? { step: 'trash', path: step.target.path }
      : { step: 'delete', path: step.target.path };
  }
  return {
    step: 'run',
    command: commandLine(step.argv),
    effect: step.effect,
    path: step.effect === 'removes' ? step.target.path : null,
  };
}

function previewIn(
  requests: readonly CleanupRequest[],
  held: ReadonlyMap<string, Item>,
  mode: DeletionMode,
): CleanupPreview {
  const entries = planAll(requests, held, mode).map((planned): CleanupEntry => ({
    item: planned.request.item,
    module: planned.item?.module ?? '',
    title: planned.title,
    action: planned.action,
    steps: planned.steps.map((step) => view(step, mode)),
    size: planned.size,
    status: planned.status,
    reversible: reversible(kindsOf(planned.steps), mode),
  }));
  const totalBytes = entries
    .filter((entry) => entry.status.state === 'ready')
    .reduce((sum, entry) => sum + entry.size, 0);
  return { entries, totalBytes, mode };
}

/** `preview_cleanup`: the batch checked in both modes, with nothing touched. */
export function mockCleanupPreview(requests: readonly CleanupRequest[]): CleanupPreviews {
  const held = heldItems();
  return {
    trash: previewIn(requests, held, 'trash'),
    permanent: previewIn(requests, held, 'permanent'),
  };
}

/** What the next batch does to an item instead of cleaning it, by item id. */
const failures = new Map<string, string>();

/**
 * Makes the first step of `itemId` fail with `message` in every batch from now on, until set
 * back to null — the mock's stand-in for a port that refuses, which the dialog's report of a
 * failure is written against.
 */
export function setMockCleanupFailure(itemId: string, message: string | null): void {
  if (message === null) failures.delete(itemId);
  else failures.set(itemId, message);
}

export function resetMockCleanup(): void {
  failures.clear();
}

/**
 * `run_cleanup`: plans the requests again, runs them with `cleanup:progress` events, records
 * the batch, forgets what went, patches the Explorer's tree and starts a rediscovery.
 *
 * The two warnings come back clear: the mock has no log that can fail to be written and no
 * patch that can be dropped.
 */
export function mockCleanupRun(
  requests: readonly CleanupRequest[],
  mode: DeletionMode,
): CleanupResult {
  const held = heldItems();
  const at = new Date().toISOString();
  const planned = planAll(requests, held, mode);
  const entries: CleanupEntryOutcome[] = [];
  let freedBytes = 0;
  for (const [done, entry] of planned.entries()) {
    publish(CLEANUP_PROGRESS_EVENT, { done, total: planned.length, current: entry.title });
    const targets: string[] = [];
    const commands: string[][] = [];
    let result: EntryResult;
    if (entry.status.state === 'blocked') {
      result = { result: 'skipped', reason: entry.status.reason };
    } else {
      const failure = failures.get(entry.request.item);
      const first = entry.steps[0];
      if (failure !== undefined) {
        // The first step was attempted and refused: its target counts as touched.
        if ('target' in first) targets.push(first.target.path);
        if (first.step === 'run') commands.push(first.argv);
        const label = first.step === 'run' ? `\`${commandLine(first.argv)}\`: ` : '';
        const where = entry.steps.length > 1 ? `step 1 of ${entry.steps.length}, ` : '';
        result = { result: 'failed', message: `${where}${label}${failure}` };
      } else {
        for (const step of entry.steps) {
          if ('target' in step) targets.push(step.target.path);
          if (step.step === 'run') commands.push(step.argv);
        }
        result = { result: 'removed', bytes: entry.size };
        freedBytes += entry.size;
      }
    }
    const target = entry.item?.path ?? null;
    const own = targetsOf(entry.steps).find((candidate) => candidate.path === target);
    entries.push({
      item: entry.request.item,
      module: entry.item?.module ?? '',
      title: entry.title,
      action: entry.action,
      path: target,
      kind: own?.kind ?? null,
      targets,
      mode: reversible(kindsOf(entry.steps), mode) ? mode : 'permanent',
      commands,
      result,
    });
  }
  publish(CLEANUP_PROGRESS_EVENT, { done: planned.length, total: planned.length, current: null });
  const outcome: CleanupOutcome = { entries, freedBytes, at, mode };
  appendCleanupToLog(outcome);
  const removed = entries
    .filter((entry) => entry.result.result === 'removed')
    .map((entry) => entry.item);
  forgetItems(removed);
  // Only targets the fixture tree knows: a patch with nothing to splice moves no id, as
  // `patch_paths` over paths the tree cannot find installs nothing. The demo's sandbox is not
  // in the fixture, which is the real "the scan never saw it".
  const gone = entries
    .filter((entry) => entry.result.result === 'removed' || entry.result.result === 'failed')
    .flatMap((entry) => entry.targets)
    .map((path) => fixtureEntry(path))
    .filter((node): node is FixtureNode => node !== undefined);
  if (gone.length > 0) {
    patchTree(gone, true);
  }
  const involved = modulesOf(requests.map((request) => request.item));
  if (involved.length > 0) {
    mockModulesRefresh(involved);
  }
  return { outcome, recorded: true, treeStale: false };
}

File: apps/desktop/src-tauri/src/cleanup.rs
//! Cleanup batches: where the modules the window holds, the cleanup engine, the record and
//! the Explorer's tree meet — `actions.rs`'s twin for items instead of paths.
//!
//! A batch is always planned again here from the requests. Nothing the UI sends is trusted
//! beyond the item ids, the actions, the options and the mode.

use std::panic::{self, AssertUnwindSafe};
use std::path::{Path, PathBuf};
use std::sync::Arc;

use storage_monitor_core::action::{ActionLog, EntryResult, Limits, Mode};
use storage_monitor_core::cleanup::{self, CleanupOutcome, Progress, Request};

use crate::actions::{BatchLock, stale_after};
use crate::module_manager::{ModuleEmitter, ModuleManager};
use crate::scan_manager::ScanManager;
use crate::views::{CleanupPreviews, CleanupResult};

pub const CLEANUP_PROGRESS_EVENT: &str = "cleanup:progress";

/// How many items `cleanup_items` hands the UI when it asks for no number of its own.
pub(crate) const DEFAULT_ITEMS_LIMIT: usize = 2000;

/// Where a batch's progress goes: the Tauri app in production, a recorder in tests.
pub trait ProgressEmitter: Send + Sync {
    fn emit(&self, progress: &Progress);
}

impl<R: tauri::Runtime> ProgressEmitter for tauri::AppHandle<R> {
    fn emit(&self, progress: &Progress) {
        if let Err(err) = tauri::Emitter::emit(self, CLEANUP_PROGRESS_EVENT, progress) {
            eprintln!("cannot emit {CLEANUP_PROGRESS_EVENT}: {err}");
        }
    }
}

/// The guards a module's deletions answer to: the home folder as the only root (phase 2b
/// design, section 8). Without a home folder there is no honest root to build them from,
/// and no batch runs.
fn limits(modules: &ModuleManager) -> Result<Limits, String> {
    modules
        .home()
        .map(|home| Limits::for_home(home.to_path_buf()))
        .ok_or_else(|| "cannot determine the home folder".to_owned())
}

/// What a batch would do in each mode, with nothing touched.
pub fn preview_cleanup(
    modules: &ModuleManager,
    requests: &[Request],
) -> Result<CleanupPreviews, String> {
    let limits = limits(modules)?;
    let sys = modules.system();
    Ok(modules.with_held(|held| CleanupPreviews {
        trash: cleanup::preview(requests, held, &limits, sys, Mode::Trash),
        permanent: cleanup::preview(requests, held, &limits, sys, Mode::Permanent),
    }))
}

/// Runs a batch, records it, forgets what it removed, patches the Explorer's tree and starts
/// a rediscovery of the modules involved — and reports which of those the window has to tell
/// the user about.
///
/// The order is `run_batch`'s, for its reasons: the record before the splice, and nothing
/// after the engine that may panic out of here, because `Err` means the batch did not run.
#[allow(clippy::too_many_arguments)]
pub fn run_cleanup(
    modules: &ModuleManager,
    scans: &ScanManager,
    log: &ActionLog,
    batches: &BatchLock,
    states: Arc<dyn ModuleEmitter>,
    progress: &dyn ProgressEmitter,
    requests: Vec<Request>,
    mode: Mode,
) -> Result<CleanupResult, String> {
    let limits = limits(modules)?;
    // Held for the whole batch: both kinds of batch patch the one tree, and a splice built
    // against a generation another batch moved is dropped.
    let _queued = batches.enter();
    let outcome = modules.with_held(|held| {
        cleanup::execute(
            &requests,
            held,
            &limits,
            modules.system(),
            mode,
            &mut |step| progress.emit(&step),
        )
    });
    let recorded = match log.append_cleanup(&outcome) {
        Ok(()) => true,
        Err(err) => {
            eprintln!("cannot record the batch in {}: {err}", log.path().display());
            false
        }
    };
    // What went is off the screen at once, before the rediscovery has anything to say.
    let removed: Vec<String> = outcome
        .entries
        .iter()
        .filter(|entry| matches!(entry.result, EntryResult::Removed { .. }))
        .map(|entry| entry.item.clone())
        .collect();
    modules.forget(&removed);
    let tree_stale = stale_after(panic::catch_unwind(AssertUnwindSafe(|| {
        scans.patch_paths(&tree_paths(scans, &outcome))
    })));
    let asked: Vec<String> = requests
        .iter()
        .map(|request| request.item.clone())
        .collect();
    let involved = modules.modules_of(&asked);
    if !involved.is_empty() {
        modules.refresh(states, &involved);
    }
    Ok(CleanupResult {
        outcome,
        recorded,
        tree_stale,
    })
}

/// The targets of every entry that was removed or failed — a failure may have deleted part of
/// what it was about, as in 2a — in the spelling the scan recorded, which is the only one
/// `ScanManager::patch_paths` can find (phase 2b design, section 10).
fn tree_paths(scans: &ScanManager, outcome: &CleanupOutcome) -> Vec<PathBuf> {
    let Some(root) = scans.root() else {
        return Vec::new();
    };
    let resolved_root = root.canonicalize().unwrap_or_else(|_| root.clone());
    outcome
        .entries
        .iter()
        .filter(|entry| {
            matches!(
                entry.result,
                EntryResult::Removed { .. } | EntryResult::Failed { .. }
            )
        })
        .flat_map(|entry| &entry.targets)
        .filter_map(|target| scan_spelling(target, &root, &resolved_root))
        .collect()
}

/// `target` as the scan spelled it, or `None` when it is not under the scan's root.
///
/// A module does not spell paths the way the scan does — `git` answers `/private/tmp` for a
/// root scanned as `/tmp` — and the engine's targets are normalized besides. So the parent is
/// resolved and the resolved root's prefix swapped for the root as scanned. The remainder then
/// carries the names the disk has, which are the names the walker read: a path the tree still
/// does not know is one the scan never saw, and there is no row to patch.
fn scan_spelling(target: &Path, root: &Path, resolved_root: &Path) -> Option<PathBuf> {
    if target.starts_with(root) {
        return Some(target.to_path_buf());
    }
    let (parent, name) = (target.parent()?, target.file_name()?);
    let resolved = parent.canonicalize().ok()?.join(name);
    let rest = resolved.strip_prefix(resolved_root).ok()?;
    Some(root.join(rest))
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::os::unix::fs::symlink;
    use std::sync::Mutex;
    use std::sync::mpsc::{self, Receiver, Sender};
    use std::thread;
    use std::time::{Duration, Instant};

    use storage_monitor_core::action::{EntryStatus, LogResult};
    use storage_monitor_core::cleanup::StepView;
    use storage_monitor_core::module::Module;
    use storage_monitor_core::system::{Reply, System, TestSystem};
    use storage_monitor_module_demo::Demo;

    use super::*;
    use crate::scan_manager::StatusEmitter;
    use crate::views::{ModuleStatus, ModuleView, ScanState, ScanStatus};

    const WAIT: Duration = Duration::from_secs(10);

    struct States(Mutex<Sender<ModuleView>>);

    impl ModuleEmitter for States {
        fn emit(&self, view: &ModuleView) {
            let _ = self.0.lock().unwrap().send(view.clone());
        }
    }

    fn states() -> (Arc<dyn ModuleEmitter>, Receiver<ModuleView>) {
        let (tx, rx) = mpsc::channel();
        (Arc::new(States(Mutex::new(tx))), rx)
    }

    #[derive(Default)]
    struct Steps(Mutex<Vec<Progress>>);

    impl ProgressEmitter for Steps {
        fn emit(&self, progress: &Progress) {
            self.0.lock().unwrap().push(progress.clone());
        }
    }

    struct Silent;

    impl StatusEmitter for Silent {
        fn emit(&self, _event: &str, _status: &ScanStatus) {}
    }

    /// The demo module over a temporary home, discovered: its sandbox seeded where the app
    /// keeps it, `rm` and `touch` scripted to do what the real ones would, through the port.
    struct World {
        sys: Arc<TestSystem>,
        modules: ModuleManager,
        /// Outside the home, where no batch can delete the record and no scan can see it.
        data: tempfile::TempDir,
    }

    impl World {
        fn new() -> Self {
            let sys = Arc::new(TestSystem::new());
            sys.install("rm");
            sys.install("touch");
            let modules = ModuleManager::new(
                vec![Box::new(Demo) as Box<dyn Module>],
                Arc::clone(&sys) as Arc<dyn System>,
                Some(sys.root().to_path_buf()),
                sys.root()
                    .join("Library/Application Support/storage-monitor"),
            );
            let world = Self {
                sys,
                modules,
                data: tempfile::tempdir().unwrap(),
            };
            world.discover();
            world
        }

        /// Refreshes the demo and waits for the answer.
        fn discover(&self) {
            let (emitter, rx) = states();
            self.modules.refresh(emitter, &[]);
            wait_for_ready(&rx);
        }

        fn path_of(&self, title: &str) -> PathBuf {
            let (items, _) = self.modules.items(100);
            items
                .into_iter()
                .find(|item| item.title == title)
                .and_then(|item| item.path)
                .unwrap_or_else(|| panic!("no {title}"))
        }

        /// Scripts the housekeeping of a folder batch and the `rm` of an object, both doing
        /// what the real tools would.
        fn script_tools(&self) {
            let sandbox = self.path_of("build-cache").parent().unwrap().to_path_buf();
            self.sys.script(
                "touch",
                &[sandbox.join(".last-cleanup").as_os_str()],
                Reply::ok(),
            );
            let old = self.path_of("old.object");
            let port = Arc::clone(&self.sys);
            let doomed = old.clone();
            self.sys.script(
                "rm",
                &[old.as_os_str()],
                Reply::ok().then(move || port.remove(&doomed).unwrap()),
            );
        }

        fn log(&self) -> ActionLog {
            ActionLog::new(self.data.path().join("actions.jsonl"))
        }

        fn scans(&self) -> ScanManager {
            ScanManager::with_snapshots_dir(self.data.path().join("snapshots"))
        }

        fn run(&self, scans: &ScanManager, requests: Vec<Request>, mode: Mode) -> CleanupResult {
            let (emitter, _rx) = states();
            run_cleanup(
                &self.modules,
                scans,
                &self.log(),
                &BatchLock::default(),
                emitter,
                &Steps::default(),
                requests,
                mode,
            )
            .expect("the batch runs")
        }
    }

    fn wait_for_ready(rx: &Receiver<ModuleView>) -> ModuleView {
        let deadline = Instant::now() + WAIT;
        loop {
            let view = rx
                .recv_timeout(deadline.saturating_duration_since(Instant::now()))
                .expect("the module answers");
            if view.status != ModuleStatus::Discovering {
                assert_eq!(view.status, ModuleStatus::Ready, "{view:?}");
                return view;
            }
        }
    }

    fn scan(scans: &ScanManager, root: &Path) {
        scans
            .start(Arc::new(Silent), root.to_path_buf())
            .expect("the scan starts");
        let deadline = Instant::now() + WAIT;
        while scans.status().state == ScanState::Running {
            assert!(Instant::now() < deadline, "the scan did not finish");
            thread::sleep(Duration::from_millis(5));
        }
        assert_eq!(scans.status().state, ScanState::Done);
    }

    fn in_tree(scans: &ScanManager, path: &Path) -> bool {
        scans
            .with_result(|result, _| result.tree.find(path).is_some())
            .expect("a scan result")
    }

    fn request(item: &str, action: &str) -> Request {
        Request {
            item: format!("demo:{item}"),
            action: action.to_owned(),
            options: Vec::new(),
        }
    }

    fn two() -> Vec<Request> {
        vec![
            request("build-cache", "delete"),
            request("old.object", "remove"),
        ]
    }

    #[test]
    fn preview_answers_both_modes_aligned() {
        let world = World::new();
        let previews = preview_cleanup(&world.modules, &two()).unwrap();
        let (trash, permanent) = (&previews.trash, &previews.permanent);
        assert_eq!(trash.entries.len(), 2);
        assert_eq!(permanent.entries.len(), 2);
        assert_eq!(trash.entries[0].item, permanent.entries[0].item);
        // The folder: two steps in Trash mode, one in Permanent.
        assert_eq!(trash.entries[0].steps.len(), 2);
        assert!(matches!(trash.entries[0].steps[0], StepView::Trash { .. }));
        assert!(trash.entries[0].reversible);
        assert_eq!(permanent.entries[0].steps.len(), 1);
        assert!(matches!(
            permanent.entries[0].steps[0],
            StepView::Delete { .. }
        ));
        // The object: the same command in both, and never undone by the Trash.
        assert!(!trash.entries[1].reversible);
        assert_eq!(trash.entries[1].steps, permanent.entries[1].steps);
        assert!(
            trash
                .entries
                .iter()
                .all(|entry| entry.status == EntryStatus::Ready),
            "{trash:?}"
        );
        assert!(world.sys.ran().is_empty(), "a preview runs nothing");
    }

    #[test]
    fn a_batch_cleans_records_forgets_and_rediscovers() {
        let world = World::new();
        world.script_tools();
        let (build, old) = (world.path_of("build-cache"), world.path_of("old.object"));
        let (emitter, rx) = states();
        let result = run_cleanup(
            &world.modules,
            &world.scans(),
            &world.log(),
            &BatchLock::default(),
            emitter,
            &Steps::default(),
            two(),
            Mode::Trash,
        )
        .unwrap();

        let results: Vec<&EntryResult> = result.outcome.entries.iter().map(|e| &e.result).collect();
        assert!(
            results
                .iter()
                .all(|r| matches!(r, EntryResult::Removed { .. })),
            "{results:?}"
        );
        assert!(world.sys.trash_dir().join("build-cache").is_dir());
        assert!(!build.exists() && !old.exists());
        assert!(result.recorded);
        assert!(!result.tree_stale);

        let tail = world.log().tail(10).unwrap();
        assert_eq!(tail.entries.len(), 2);
        let source = tail.entries[0].source.as_ref().expect("a cleanup line");
        assert_eq!(source.module, "demo");
        assert_eq!(tail.entries[0].result, LogResult::Removed);
        assert_eq!(
            tail.entries[0].mode,
            Mode::Permanent,
            "the object left for good"
        );
        assert_eq!(
            tail.entries[1].mode,
            Mode::Trash,
            "the folder is in the Trash"
        );

        // Gone from the screen at once, and a rediscovery of the demo is on its way.
        let titles: Vec<String> = world
            .modules
            .items(100)
            .0
            .into_iter()
            .map(|i| i.title)
            .collect();
        assert!(!titles.contains(&"build-cache".to_owned()), "{titles:?}");
        assert!(!titles.contains(&"old.object".to_owned()), "{titles:?}");
        let ready = wait_for_ready(&rx);
        assert_eq!(ready.id, "demo");
        assert_eq!(ready.item_count, 3);
    }

    #[test]
    fn a_batch_patches_the_explorer_tree() {
        let world = World::new();
        world.script_tools();
        let scans = world.scans();
        scan(&scans, world.sys.root());
        let build = world.path_of("build-cache");
        assert!(in_tree(&scans, &build), "the scan saw it");
        let before = scans.status().bytes;

        let result = world.run(&scans, vec![request("build-cache", "delete")], Mode::Trash);
        assert!(!result.tree_stale);
        assert!(!in_tree(&scans, &build), "the row is gone");
        assert!(scans.status().bytes < before, "and the root shrank");
    }

    #[test]
    fn a_target_spelled_otherwise_is_moved_into_the_scan_spelling() {
        let world = World::new();
        world.script_tools();
        // The home scanned through a symlink: the tree spells every path through the link,
        // while the demo's paths are the resolved ones.
        let links = tempfile::tempdir().unwrap();
        let through = links.path().join("home");
        symlink(world.sys.root(), &through).unwrap();
        let scans = world.scans();
        scan(&scans, &through);
        let build = world.path_of("build-cache");
        let spelled = through.join(build.strip_prefix(world.sys.root()).unwrap());
        assert!(in_tree(&scans, &spelled), "the scan's spelling");
        assert!(!in_tree(&scans, &build), "which is not the module's");

        let result = world.run(&scans, vec![request("build-cache", "delete")], Mode::Trash);
        assert!(!result.tree_stale);
        assert!(!in_tree(&scans, &spelled), "patched all the same");
    }

    #[test]
    fn a_target_the_scan_never_saw_is_not_stale() {
        let world = World::new();
        let scans = world.scans();
        // Scanned before the sandbox existed.
        let sandbox = world.path_of("build-cache").parent().unwrap().to_path_buf();
        world.sys.remove(&sandbox).unwrap();
        scan(&scans, world.sys.root());
        world.discover();
        world.script_tools();

        let result = world.run(&scans, vec![request("build-cache", "delete")], Mode::Trash);
        assert!(matches!(
            result.outcome.entries[0].result,
            EntryResult::Removed { .. }
        ));
        assert!(
            !result.tree_stale,
            "no row was left behind, so nothing to warn about"
        );
    }

    #[test]
    fn progress_is_emitted_per_entry() {
        let world = World::new();
        world.script_tools();
        let steps = Steps::default();
        let (emitter, _rx) = states();
        run_cleanup(
            &world.modules,
            &world.scans(),
            &world.log(),
            &BatchLock::default(),
            emitter,
            &steps,
            two(),
            Mode::Trash,
        )
        .unwrap();
        let seen = steps.0.into_inner().unwrap();
        assert_eq!(
            seen.iter().map(|p| (p.done, p.total)).collect::<Vec<_>>(),
            vec![(0, 2), (1, 2), (2, 2)]
        );
        assert_eq!(seen[0].current.as_deref(), Some("build-cache"));
        assert_eq!(seen[2].current, None);
    }

    #[test]
    fn cleanup_and_explorer_batches_share_the_queue() {
        let world = World::new();
        world.script_tools();
        let batches = Arc::new(BatchLock::default());

        /// Checks, while the batch runs, that the queue both kinds of batch wait in is held.
        struct Watching(Arc<BatchLock>, Mutex<Vec<bool>>);

        impl ProgressEmitter for Watching {
            fn emit(&self, _progress: &Progress) {
                self.1.lock().unwrap().push(self.0.is_held());
            }
        }

        let watching = Watching(Arc::clone(&batches), Mutex::default());
        let (emitter, _rx) = states();
        run_cleanup(
            &world.modules,
            &world.scans(),
            &world.log(),
            &batches,
            emitter,
            &watching,
            two(),
            Mode::Trash,
        )
        .unwrap();
        let held = watching.1.into_inner().unwrap();
        assert!(
            !held.is_empty() && held.iter().all(|held| *held),
            "{held:?}"
        );
        assert!(!batches.is_held(), "and let go of afterwards");
    }

    #[test]
    fn an_unrecorded_batch_still_reports_its_outcome() {
        let world = World::new();
        world.script_tools();
        // A folder where the log's own folder should be: nothing can be appended.
        let blocked = world.data.path().join("blocked");
        fs::write(&blocked, b"a file").unwrap();
        let (emitter, _rx) = states();
        let result = run_cleanup(
            &world.modules,
            &world.scans(),
            &ActionLog::new(blocked.join("actions.jsonl")),
            &BatchLock::default(),
            emitter,
            &Steps::default(),
            two(),
            Mode::Trash,
        )
        .unwrap();
        assert!(!result.recorded);
        assert!(matches!(
            result.outcome.entries[0].result,
            EntryResult::Removed { .. }
        ));
    }

    #[test]
    fn without_a_home_folder_no_batch_runs() {
        let sys = Arc::new(TestSystem::new());
        let modules = ModuleManager::new(
            vec![Box::new(Demo) as Box<dyn Module>],
            Arc::clone(&sys) as Arc<dyn System>,
            None,
            sys.root().to_path_buf(),
        );
        assert!(preview_cleanup(&modules, &two()).is_err());
        let data = tempfile::tempdir().unwrap();
        let (emitter, _rx) = states();
        let err = run_cleanup(
            &modules,
            &ScanManager::with_snapshots_dir(data.path().join("snapshots")),
            &ActionLog::new(data.path().join("actions.jsonl")),
            &BatchLock::default(),
            emitter,
            &Steps::default(),
            two(),
            Mode::Trash,
        )
        .unwrap_err();
        assert_eq!(err, "cannot determine the home folder");
        assert!(
            !data.path().join("actions.jsonl").exists(),
            "nothing ran, nothing recorded"
        );
    }

    #[test]
    fn scan_spelling_leaves_a_path_outside_the_scan_alone() {
        let dir = tempfile::tempdir().unwrap();
        let root = dir.path().canonicalize().unwrap();
        let outside = tempfile::tempdir().unwrap();
        let target = outside.path().canonicalize().unwrap().join("x");
        assert_eq!(scan_spelling(&target, &root, &root), None);
        assert_eq!(
            scan_spelling(&root.join("a/b"), &root, &root),
            Some(root.join("a/b")),
            "the scan's own spelling passes as it is"
        );
    }
}
